Source/WebCore/page/Page.cpp

```diff
@@ -1730,6 +1732,15 @@ void Page::updateRendering()
         document.serviceRequestAnimationFrameCallbacks();
     });
 
+    runProcessingStep(RenderingUpdateStep::FocusFixupRule, [] (Document& document) {
+        if (auto* activeElement = document.activeElement()) {
+            if (auto* root = activeElement->shadowRoot(); root && root->activeElement())
```

Comment:
Ditto.

Reply:
This doesn't seem right. This root's activeElement may in turn be inside another shadow root.

Reply:
We clearly need more tests. One where there are multiple nested shadow trees.

Reply:
This doesn't seem right. This root's activeElement may in turn be inside another shadow root.

We clearly need more tests. One where there are multiple nested shadow trees.

Reply:
Also, we should probably start from document.focusedElement() and walk the tree upwards instead. Instead of keep walking up the tree multiple times in `TreeScope::focusedElementInScope()`

Reply:
That may be the source of the test failures I'm seeing.